chalk-solve/src/solve.rs

```diff
@@ -128,6 +128,35 @@ impl Solver {
         self.forest.solve(&ops, goal)
     }
 
+    /// Attempts to solve the given goal, which must be in canonical
+    /// form. Provides multiple solutions to function `f`.  This will do
+    /// only as much work towards `goal` as it has to (and that work
+    /// is cached for future attempts).
+    ///
+    /// # Parameters
+    ///
+    /// - `program` -- defines the program clauses in scope.
+    ///   - **Important:** You must supply the same set of program clauses
+    ///     each time you invoke `solve`, as otherwise the cached data may be
+    ///     invalid.
+    /// - `goal` the goal to solve
+    /// - `f` -- function to proceed solution. New solutions will be generated
```

Comment:
It's not obvious what the parameters to `f` represent, can you comment them here?

Reply:
I've added comment about parameters and return value.